spec/java/util/HashSet.main.lsl

```diff
@@ -0,0 +1,633 @@
+        val iter: Iterator = action CALL_METHOD(c, "iterator", []);
+        val lengthBeforeRemoving: int = this.length;
+        var i: int = 0;
+
+        if (this.length > otherSize)
+        {
+            action LOOP_WHILE(
+                action CALL_METHOD(iter, "hasNext", []),
+                removeAllElements_loop_direct(iter)
+            );
+        }
+        else
+        {
+            val visitedKeys: map<Object, Object> = action MAP_NEW();
+            action LOOP_WHILE(
+                i < this.length,
+                _removeAllElements_loop_indirect(i, c, visitedKeys)
+            );
+        }
+
+        _checkForComodification(expectedModCount);
+        this.modCount += 1;
+        // If length changed, it means that at least one element was deleted
+        result = lengthBeforeRemoving != this.length;
+    }
+
+
+    @Phantom proc removeAllElements_loop_direct (iter: Iterator): void
+    {
+        val key: Object = action CALL_METHOD(iter, "next", []);
+        val isKeyExist: boolean = action MAP_HAS_KEY(this.storage, key);
+
+        if (isKeyExist)
+        {
+            action MAP_REMOVE(this.storage, key);
+            this.length -= 1;
+        }
+    }
+
+
+    @Phantom proc _removeAllElements_loop_indirect (i: int, c: Collection, visitedKeys: map<Object, Object>): void
+    {
+        val key: Object = _generateKey(visitedKeys);
+
+        val isCollectionContainsKey: boolean = action CALL_METHOD(c, "contains", [key]);
+
+        if (isCollectionContainsKey)
+        {
+            action MAP_REMOVE(this.storage, key);
+            this.length -= 1;
+        }
+
+        action MAP_SET(visitedKeys, key, HASHSET_VALUE);
+        i += 1;
+    }
+
+
+    fun *.toArray (@target self: HashSet): array<Object>
+    {
+        val expectedModCount: int = this.modCount;
+        val size: int = this.length;
+        val visitedKeys: map<Object, Object> = action MAP_NEW();
+        val resultArray: array<Object> = action ARRAY_NEW("java.lang.Object", this.length);
+        var i: int = 0;
+
+        action LOOP_FOR(
+            i, 0, size, +1,
+            toArray_loop(i, visitedKeys, resultArray)
+        );
+
+        _checkForComodification(expectedModCount);
+        result = resultArray;
+    }
+
+
+    @Phantom proc toArray_loop(i: int, visitedKeys: map<Object, Object>, resultArray: array<Object>): void
+    {
+        val key: Object = _generateKey(visitedKeys);
+
+        resultArray[i] = key;
+
+        action MAP_SET(visitedKeys, key, HASHSET_VALUE);
+    }
+
+
+    fun *.toArray (@target self: HashSet, a: array<Object>): array<Object>
+    {
+        val expectedModCount: int = this.modCount;
+        val aLen: int = action ARRAY_SIZE(a);
+        val size: int = this.length;
+        val visitedKeys: map<Object, Object> = action MAP_NEW();
+        var resultArray: array<Object> = action ARRAY_NEW("java.lang.Object", this.length);
+        var i: int = 0;
+
+        if (aLen < size)
```

Comment:
This can be simplified similar to [that in `ArrayList`](https://github.com/vpa-research/jsl-spec/blob/patches-mikhail/spec/java/util/ArrayList.main.lsl#L865).

Reply:
Fix